src/provider.tsx

```diff
@@ -0,0 +1,160 @@
+
+import { auth0Loaded, Auth0ProviderAction, handleRedirectCallbackAction, loginWithPopupAction } from "./actions";
+import { useLog } from "./hooks";
+import { auth0ProviderStateReducer } from "./reducer";
+import {
+  Auth0ProviderContext,
+  Auth0ProviderProps,
+  Auth0ProviderState,
+  isHandlingRedirectState,
+  isLoadedState,
+  isLoadingState,
+  LoadingState,
+  RouteState,
+  TokenUser,
+} from "./types";
+
+const InitialReducerState: LoadingState = {
+  loading: true,
+  isAuthenticated: false,
+  popupOpen: false,
+  user: undefined,
+  auth0Client: undefined,
+  handlingRedirect: false,
+};
+
+const buildHandleRedirectCallback = (
+  dispatch: Dispatch<Auth0ProviderAction>,
+  log: typeof console.log,
+  auth0Client: Auth0Client,
+) => async (url?: string): Promise<RedirectLoginResult> => {
+  dispatch(handleRedirectCallbackAction.started());
+
+  const response = await auth0Client.handleRedirectCallback(url);
+  const user = await auth0Client.getUser();
+  log("handleRedirectCallback(): got response:", response);
+  log("handleRedirectCallback(): got user:", user);
+
+  dispatch(handleRedirectCallbackAction.done({ result: { user } }));
+
+  return response;
+};
+
+const buildLoginWithPopup = (dispatch: Dispatch<Auth0ProviderAction>, auth0Client: Auth0Client) => async (
+  options?: PopupLoginOptions,
+  config?: PopupConfigOptions,
+): Promise<void> => {
+  dispatch(loginWithPopupAction.started());
+
+  try {
+    await auth0Client.loginWithPopup(options, config);
+  } catch (error) {
+    console.error(error);
+  }
+
+  const user = await auth0Client.getUser();
+
+  dispatch(loginWithPopupAction.done({ result: { user } }));
+};
+
+export const Auth0Provider: React.FC<Auth0ProviderProps> = ({
+  children,
+  history,
+  enableDebugLogging,
+  renderLoading,
+  ...initOptions
+}) => {
+  const log = useLog(enableDebugLogging);
+
+  const [state, dispatch] = useReducer<Reducer<Auth0ProviderState, Auth0ProviderAction>>(
+    auth0ProviderStateReducer,
+    InitialReducerState,
+  );
+
+  const location = useLocation();
+
+  const { loading: initLoading, value: initValue } = useAsync(async () => {
+    log("Beginning auth0 initialization");
+
+    dispatch(auth0Loaded.started());
+
+    const auth0Client = await createAuth0Client(initOptions);
+
+    const handleRedirectCallback = buildHandleRedirectCallback(dispatch, log, auth0Client);
+
+    const loginWithPopup = buildLoginWithPopup(dispatch, auth0Client);
+
+    // if (location.search?.includes("code=") && location.search?.includes("state=")) {
+    //   log("Handling redirect");
+    //   const { appState } = await handleRedirectCallback();
+    //   onRedirectCallback(appState);
+    // }
+
+    const isAuthenticated = await auth0Client.isAuthenticated();
+
+    const user: TokenUser | undefined = (isAuthenticated && (await auth0Client.getUser())) || undefined;
+
+    return {
+      user,
+      auth0Client,
+      handleRedirectCallback,
+      loginWithPopup,
+    };
+  }, []);
```

Comment:
this should depend on `dispatch` and `log` right?

Reply:
Yes.  I originally had those in, and was dealing with some odd re-render behavior which is now resolved, and I neglected to return them.   Good catch -- the eslint checker doesn't catch deps on the `react-use` hooks :)